### PR TITLE
Pill Component

### PR DESCRIPTION
**Pull request by graygilmore**
This PR introduces a new "pill" component. This component is used to display the state of something through six distinct colors:

- Complete
- Active
- Inactive
- Pending
- Warning
- Error

This should make scanning tables significantly easier and provide a slightly more modern feel to the admin (looks great with the new tables in https://github.com/solidusio/solidus/pull/2092).

## TODO

- [x] fix test errors

## Style Guide

screen shot 2017-07-28 at 11 32 11 am

## Orders (before)

screen shot 2017-07-28 at 11 27 07 am

## Orders (after)

screen shot 2017-07-28 at 11 26 11 am

## Promotions (before)

screen shot 2017-07-28 at 11 27 25 am

## Promotions (after)

screen shot 2017-07-28 at 11 27 48 am

### CODE REVIEW

**Comment by Mandily** [comment]
I like the content of what you're trying to say in the writeup in the style guide, but could we write it in a clearer way? Maybe something like this: 

"Pill styling adds a second, easily scannable way to recognize status. They're most commonly used in, but not restricted to tables."

**Comment by tvdeyen** [comment]
I like that! 😍 

Maybe it's just me, but shouldn't `complete` be green? I assume shop owners love to see lots of green in their orders table.

File: backend/app/assets/stylesheets/spree/backend/components/_pills.scss [hunk 1]
```diff
@@ -0,0 +1,43 @@
+.pill {
+  display: inline-block;
+  padding: 2px 10px;
+  border-radius: 10px;
+  font-size: 11px;
```
Comment (tvdeyen) [inline backend/app/assets/stylesheets/spree/backend/components/_pills.scss]: The `font-size` should have a relative unit.
Also the paddings.

File: backend/app/views/spree/admin/promotions/index.html.erb [hunk 2]
```diff
@@ -85,11 +77,26 @@
...
+          </td>
+          <td class="align-center">
+            <%= promotion.usage_count %>
+          </td>
+          <td class="align-center">
+            <%= promotion.starts_at.to_date.to_s(:short_date) if promotion.starts_at %>
+          </td>
+          <td class="align-center">
+            <%= promotion.expires_at.to_date.to_s(:short_date) if promotion.expires_at %>
+          </td>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/promotions/index.html.erb]: You are changing the table far more than just adding the pill component. Could you please do this in a separate commit, or not even PR?

File: core/config/locales/en.yml [hunk 3]
```diff
@@ -170,13 +170,14 @@ en:
...
+        expires_at: End
         name: Name
         path: Path
         per_code_usage_limit: Per Code Usage Limit
         promotion_uses: Promotion Uses
-        starts_at: Start date
-        usage_limit: Overall Usage Limit
+        starts_at: Start
+        status: Status
+        usage_limit: Usage Limit
```
Comment (tvdeyen) [inline core/config/locales/en.yml]: These changes do not belong in this PR IMO. At least it should be a separate commit.

File: backend/app/helpers/spree/admin/reimbursements_helper.rb [hunk 4]
```diff
@@ -1,14 +0,0 @@
...
-        case reimbursement.reimbursement_status
-        when 'reimbursed' then 'success'
-        when 'pending' then 'notice'
-        when 'errored' then 'error'
-        else raise "unknown reimbursement status: #{reimbursement.reimbursement_status}"
-        end
-      end
-    end
-  end
-end
```
Comment (tvdeyen) [inline backend/app/helpers/spree/admin/reimbursements_helper.rb]: Could we move this removal in a separate commit after this commit?

File: backend/app/helpers/spree/admin/stock_transfers_helper.rb [hunk 5]
```diff
@@ -17,10 +17,6 @@ def stock_transfer_edit_or_ship_path(stock_transfer)
         end
       end
 
-      def stock_transfer_status(stock_transfer)
-        stock_transfer.closed? ? Spree.t(:closed) : Spree.t(:open)
-      end
```
Comment (tvdeyen) [inline backend/app/helpers/spree/admin/stock_transfers_helper.rb]: this as well, please

File: backend/app/views/spree/admin/cancellations/index.html.erb [hunk 6]
```diff
@@ -34,7 +34,15 @@
...
           </td>
           <td class="inventory-unit-quantity align-center">1</td>
-          <td class="inventory-unit-state align-center"><%= inventory_unit.state %></td>
+          <td class="inventory-unit-state align-center">
+            <span class="pill pill-<%= inventory_unit.state %>">
+              <%= Spree.t(
+                inventory_unit.state,
+                scope: :inventory_states,
+                default: inventory_unit.state.titleize
+              ).titleize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/cancellations/index.html.erb]: I like that this is now translatable, but please do not `titleize` this. This is what should be done in the translation itself, upper and lower case rules are very specific to languages.

Also, this should be a separate commit.

File: backend/app/views/spree/admin/customer_returns/_reimbursements_table.html.erb [hunk 7]
```diff
@@ -20,8 +20,8 @@
         </td>
         <td><%= reimbursement.display_total %></td>
         <td>
-          <span class="state <%= reimbursement_status_color(reimbursement) %>">
-            <%= reimbursement.reimbursement_status %>
+          <span class="pill pill-<%= reimbursement.reimbursement_status %>">
+            <%= reimbursement.reimbursement_status.titleize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/customer_returns/_reimbursements_table.html.erb]: Please do not transform strings like this, pass it through `I18n` or just leave it like it is.

File: backend/app/views/spree/admin/customer_returns/_return_item_decision.html.erb [hunk 8]
```diff
@@ -39,7 +39,9 @@
           <%= return_item.acceptance_status_errors %>
         </td>
         <td class-"align-center">
-          <%= return_item.reception_status.humanize %>
+          <span class="pill pill-<%= return_item.reception_status %>">
+            <%= return_item.reception_status.humanize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/customer_returns/_return_item_decision.html.erb]: We should only use `humanize` if we pass it through `I18n` before and humanize the fallback translation.

```ruby
Spree.t(return_item.reception_status, 
  scope: 'return_item_reception_statuses', 
  default: return_item.reception_status.humanize)
```

File: backend/app/views/spree/admin/customer_returns/_return_item_selection.html.erb [hunk 9]
```diff
@@ -38,7 +38,9 @@
           <%= return_item.display_amount %>
         </td>
         <td class="align-center">
-          <%= return_item.inventory_unit.state %>
+          <span class="pill pill-<%= return_item.inventory_unit.state %>">
+            <%= return_item.inventory_unit.state.titleize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/customer_returns/_return_item_selection.html.erb]: Same comment as above. I suggest to make a separate commit (not even a PR) first that changes the presentation via `I18n`.

File: backend/app/views/spree/admin/orders/confirm/_payments.html.erb [hunk 10]
```diff
@@ -30,8 +30,12 @@
...
           </td>
           <td class="align-center">
-            <span class="state <%= payment.state %>">
-              <%= Spree.t(payment.state, scope: :payment_states, default: payment.state.capitalize) %>
+            <span class="pill pill-<%= payment.state %>">
+              <%= Spree.t(
+                payment.state,
+                scope: :payment_states,
+                default: payment.state
+              ).titleize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/orders/confirm/_payments.html.erb]: Ditto

File: backend/app/views/spree/admin/orders/index.html.erb [hunk 11]
```diff
@@ -150,11 +150,27 @@
       <tr data-hook="admin_orders_index_rows" class="state-<%= order.state.downcase %> <%= cycle('odd', 'even') %>">
         <td class="align-center"><%= l (@show_only_completed ? order.completed_at : order.created_at).to_date %></td>
         <td class="align-center"><%= link_to order.number, edit_admin_order_path(order) %></td>
-        <td class="align-center"><span class="state <%= order.state.downcase %>"><%= Spree.t("order_state.#{order.state.downcase}") %></span></td>
-        <td class="align-center"><span class="state <%= order.payment_state %>"><%= link_to Spree.t("payment_states.#{order.payment_state}"), admin_order_payments_path(order) if order.payment_state %></span></td>
-          <% if Spree::Order.checkout_step_names.include?(:delivery) %>
-            <td class="align-center"><span class="state <%= order.shipment_state %>"><%= Spree.t("shipment_states.#{order.shipment_state}") if order.shipment_state %></span></td>
+        <td class="align-center">
+          <span class="pill pill-<%= order.state.downcase %>">
+            <%= Spree.t("order_state.#{order.state.downcase}").titleize %>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/orders/index.html.erb]: Same comment because of titleization applies here

File: backend/app/views/spree/admin/users/items.html.erb [hunk 12]
```diff
@@ -52,15 +50,6 @@
...
               <td class="align-center item-total"><%= item.money.to_html %></td>
-              <td class="align-center order-state">
-                <div class="state <%= order.state.downcase %>"><%= Spree.t("order_state.#{order.state.downcase}") %></div>
-                <% if order.payment_state %>
-                  <div class="state <%= order.payment_state %>"><%= link_to Spree.t("payment_states.#{order.payment_state}"), admin_order_payments_path(order) %></div>
-                <% end %>
-                <% if Spree::Order.checkout_step_names.include?(:delivery) && order.shipment_state %>
-                  <div class="state <%= order.shipment_state %>"><%= Spree.t("shipment_states.#{order.shipment_state}") %></div>
-                <% end %>
-              </td>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/users/items.html.erb]: Should be a separate commit as well

**Comment by tvdeyen** [comment]
The build errors are actually related to the reimbursements helper removal.

**Comment by Mandily** [comment]
re: "Shouldn't complete be green?" 

Green is usually associated with an action - like "go" or "active" in this case. In the case of grey being used for complete, it means that a thing is done and you don't have to worry about doing anything else for it. The colours stand out more than the greys because they're the things still in flux that an admin might want to keep an eye on. The neutral things are the ones they don't have to worry about.

The caveat here being that most if not all of my colour theory is based on North American norms. Maybe these colours mean different things to you.

Make sense?

**Comment by graygilmore** [comment]
> Maybe it's just me, but shouldn't complete be green? I assume shop owners love to see lots of green in their orders table.

No. Green in this case specifies something that is `active`. It gives the feeling of movement and that something is currently happening. A `complete` state is something that needs no further action and is finished, and the grey-blue color better helps signal that.

File: backend/app/assets/stylesheets/spree/backend/components/_pills.scss (hunk 1 above)
Comment (graygilmore) [inline backend/app/assets/stylesheets/spree/backend/components/_pills.scss]: Any particular reason?

File: backend/app/views/spree/admin/cancellations/index.html.erb (hunk 6 above)
Comment (graygilmore) [inline backend/app/views/spree/admin/cancellations/index.html.erb]: 👍 will make changes to the translations and then `downcase` or somehow slugify for the class names.

File: backend/app/assets/stylesheets/spree/backend/components/_pills.scss (hunk 1 above)
Comment (tvdeyen) [inline backend/app/assets/stylesheets/spree/backend/components/_pills.scss]: paddings not passings. (Sorry iOS)

File: backend/app/views/spree/admin/cancellations/index.html.erb (hunk 6 above)
Comment (tvdeyen) [inline backend/app/views/spree/admin/cancellations/index.html.erb]: > we're not i18n-ing those!

😆

**Comment by tvdeyen** [comment]
> Re: Re: Green

This makes total sense. Thanks for the explanation.

File: backend/app/assets/stylesheets/spree/backend/components/_pills.scss (hunk 1 above)
Comment (graygilmore) [inline backend/app/assets/stylesheets/spree/backend/components/_pills.scss]: I'm more curious about who the perspective is here. This is a very specific component that I have sized specifically. If users viewing the admin bump up their font size this will scale along with it.

If we decide to change the root font size of the admin we'll need to go through components one by one and figure out what needs to change. I'd normally be happy to use `rem` but our admin root font size is `13px` which I can see changing in the future but I wouldn't want this component to change.

Comment (tvdeyen) [inline backend/app/assets/stylesheets/spree/backend/components/_pills.scss]: I like us to move forward to relative units all over the place. That’s the way all frontend frameworks went for a reason. 

I see the problem with the small base font size we have. So, I won’t insist to make this change right now. 

But, could you please make this font size a variable then?

**Comment by graygilmore** [comment]
Wooo tests are green. @tvdeyen I think I've addressed all of your other concerns. I went a little overly verbose with the number of commits but that should make things a little cleaner to understand in the future.

File: backend/app/views/spree/admin/shared/_order_summary.html.erb [hunk 13]
```diff
@@ -1,7 +1,15 @@
 <header id="order_tab_summary" data-hook>
   <dl class="additional-info">
     <dt class="order-state" id="order_status" data-hook><%= Spree.t(:status) %>:</dt>
-    <dd class="order-state"><span class="state <%= @order.state %>"><%= Spree.t(@order.state, scope: :order_state) %></span></dd>
+    <dd class="order-state">
+      <span class="pill pill-<%= @order.state %>">
+        <%= Spree.t(
+          @order.state,
+          scope: :order_state,
+          default: @order.state.humanize
```
Comment (jhawthorn) [inline backend/app/views/spree/admin/shared/_order_summary.html.erb]: I'd rather not use `humanize` as a default. We want to see where we are missing translations (and avoid the cost of calling the inflector)

File: core/config/locales/en.yml [hunk 14]
```diff
@@ -1645,6 +1645,7 @@ en:
       pending: Pending
       processing: Processing
       void: Void
+      invalid: Invalid
```
Comment (graygilmore) [inline core/config/locales/en.yml]: Nice! Thanks for that.

**Comment by Eth3rnit3** [comment]
capture d ecran 2017-10-18 a 09 49 22
I have this error message after the update to 2.4 do you have any idea where it might come from?
server-side : 
`ActionView::Template::Error (".state.sold" failed to @extend ".pill-sold".
The selector ".pill-sold" was not found.
Use "@extend .pill-sold !optional" if the extend should be able to fail.
):
    4: <meta content="width=device-width, initial-scale=1.0, maximum-scale=1" name="viewport">
    5: <%== meta_data_tags %>
    6: <%= canonical_tag(current_store.url) %>
    7: <%= favicon_link_tag 'favicon.ico' %>
    8: <%= stylesheet_link_tag 'spree/frontend/all', media: 'screen' %>
    9: <%= csrf_meta_tags %>
   10: <%= javascript_include_tag 'spree/frontend/all' %>`

**Comment by mamhoff** [comment]
@Eth3rnit3 Have you overridden any style sheets in your host app? Compare them with the ones in Solidus 2.4. You're probably not requiring the pills SCSS file.

**Comment by Eth3rnit3** [comment]
yes only the file variables_override. scss I'm going to compare it right now and I'm telling you

**Comment by Eth3rnit3** [comment]
no unfortunately the file is empty in solidus

**Comment by Eth3rnit3** [comment]
you're right, if I add the line: 

> @import 'spree/backend/components/pills'
to my file, it works; -) Thank you

**Comment by Eth3rnit3** [comment]
not really sorry, actually. 
only a css problem that made a mistake earlier

**Comment by Eth3rnit3** [comment]
actually the problem was that I overloaded the $state class in my variables_override file. thank you, maybe we'll see you soon.
